File: SIH-Project-main/server/src/api/v1/controllers/safety.controller.js
const safetyService = require('../../../services/safety.service');

const safetyController = {
  /**
   * Trigger emergency SOS alert
   * POST /api/v1/safety/sos-trigger
   */
  async triggerSos(req, res, next) {
    try {
      const { userId, userName, userPhone, location, emergencyContacts, autoDialNumber } = req.body;

      const result = await safetyService.triggerSos({
        userId: userId || req.user?.uid || 'anonymous_user',
        userName: userName || req.user?.displayName || 'Traveler',
        userPhone: userPhone || req.user?.phone || '',
        location: location || { lat: 26.9124, lng: 75.7873 },
        emergencyContacts: emergencyContacts || req.user?.emergencyContacts || [],
        autoDialNumber: autoDialNumber || '112',
      });

      return res.status(201).json({
        success: true,
        message: 'Emergency SOS alert activated and dispatched to contacts',
        data: result,
        timestamp: new Date().toISOString(),
      });
    } catch (error) {
      next(error);
    }
  },

  /**
   * Start Live Trip Location Sharing
   * POST /api/v1/safety/share-trip/start
   */
  async startTripShare(req, res, next) {
    try {
      const { userId, userName, userPhone, location, emergencyContacts, durationHours } = req.body;

      const result = await safetyService.startTripShare({
        userId: userId || req.user?.uid || 'anonymous_user',
        userName: userName || req.user?.displayName || 'Traveler',
        userPhone: userPhone || req.user?.phone || '',
        location: location || { lat: 26.9124, lng: 75.7873 },
        emergencyContacts: emergencyContacts || [],
        durationHours: durationHours || 12,
      });

      return res.status(201).json({
        success: true,
        message: 'Live trip location sharing session started',
        data: result,
        timestamp: new Date().toISOString(),
      });
    } catch (error) {
      next(error);
    }
  },

  /**
   * Periodic live location ping update
   * POST /api/v1/safety/share-trip/update
   */
  async updateLiveLocation(req, res, next) {
    try {
      const { sessionId, lat, lng, speed, battery, address } = req.body;

      if (!sessionId || lat === undefined || lng === undefined) {
        return res.status(400).json({
          success: false,
          message: 'sessionId, lat, and lng are required',
          timestamp: new Date().toISOString(),
        });
      }

      const updated = await safetyService.updateLiveLocation({
        sessionId,
        lat: parseFloat(lat),
        lng: parseFloat(lng),
        speed: speed ? parseFloat(speed) : 0,
        battery: battery ? parseInt(battery, 10) : 80,
        address,
      });

      return res.status(200).json({
        success: true,
        message: 'Live location updated successfully',
        data: updated,
        timestamp: new Date().toISOString(),
      });
    } catch (error) {
      next(error);
    }
  },

  /**
   * Stop active sharing or SOS session
   * POST /api/v1/safety/share-trip/stop
   */
  async stopSession(req, res, next) {
    try {
      const { sessionId, userId } = req.body;

      if (!sessionId) {
        return res.status(400).json({
          success: false,
          message: 'sessionId is required to stop session',
          timestamp: new Date().toISOString(),
        });
      }

      const stopped = await safetyService.stopSession({
        sessionId,
        userId: userId || req.user?.uid,
      });

      return res.status(200).json({
        success: true,
        message: 'Location sharing session stopped',
        data: stopped,
        timestamp: new Date().toISOString(),
      });
    } catch (error) {
      next(error);
    }
  },

  /**
   * Get active session by ID (used by web map viewer)
   * GET /api/v1/safety/session/:sessionId
   */
  async getSession(req, res, next) {
    try {
      const { sessionId } = req.params;
      const session = await safetyService.getSession(sessionId);

      if (!session) {
        return res.status(404).json({
          success: false,
          message: 'Safety session not found',
          timestamp: new Date().toISOString(),
        });
      }

      return res.status(200).json({
        success: true,
        message: 'Safety session retrieved',
        data: session,
        timestamp: new Date().toISOString(),
      });
    } catch (error) {
      next(error);
    }
  },
};

module.exports = safetyController;

File: SIH-Project-main/app/lib/features/location/presentation/controllers/safety_controller.dart
import 'dart:async';
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:url_launcher/url_launcher.dart';
import '../../../../core/network/api_client.dart';
import '../../auth/domain/models/emergency_contact.dart';
import '../../auth/presentation/controllers/auth_controller.dart';
import '../data/datasources/safety_remote_data_source.dart';
import '../data/repositories/safety_repository_impl.dart';
import '../domain/models/safety_models.dart';
import '../domain/repositories/safety_repository.dart';

class SafetyState {
  final bool isSosActive;
  final bool isLocationSharingActive;
  final bool isSendingSos;
  final String? activeSessionId;
  final String? publicTrackingUrl;
  final LiveLocationData currentLocation;
  final SosTriggerResult? lastSosResult;
  final String? errorMessage;

  const SafetyState({
    this.isSosActive = false,
    this.isLocationSharingActive = false,
    this.isSendingSos = false,
    this.activeSessionId,
    this.publicTrackingUrl,
    this.currentLocation = const LiveLocationData(
      lat: 26.9124,
      lng: 75.7873,
      accuracy: 8,
      speed: 0,
      battery: 88,
      address: 'Pink City, Jaipur, Rajasthan',
    ),
    this.lastSosResult,
    this.errorMessage,
  });

  SafetyState copyWith({
    bool? isSosActive,
    bool? isLocationSharingActive,
    bool? isSendingSos,
    String? activeSessionId,
    String? publicTrackingUrl,
    LiveLocationData? currentLocation,
    SosTriggerResult? lastSosResult,
    String? errorMessage,
  }) {
    return SafetyState(
      isSosActive: isSosActive ?? this.isSosActive,
      isLocationSharingActive: isLocationSharingActive ?? this.isLocationSharingActive,
      isSendingSos: isSendingSos ?? this.isSendingSos,
      activeSessionId: activeSessionId ?? this.activeSessionId,
      publicTrackingUrl: publicTrackingUrl ?? this.publicTrackingUrl,
      currentLocation: currentLocation ?? this.currentLocation,
      lastSosResult: lastSosResult ?? this.lastSosResult,
      errorMessage: errorMessage,
    );
  }
}

final safetyRepositoryProvider = Provider<SafetyRepository>((ref) {
  final apiClient = ApiClient();
  final remoteDataSource = SafetyRemoteDataSourceImpl(apiClient);
  return SafetyRepositoryImpl(remoteDataSource);
});

final safetyControllerProvider =
    StateNotifierProvider<SafetyNotifier, SafetyState>((ref) {
  final repo = ref.watch(safetyRepositoryProvider);
  return SafetyNotifier(repo, ref);
});

class SafetyNotifier extends StateNotifier<SafetyState> {
  final SafetyRepository _repository;
  final Ref _ref;
  Timer? _periodicLocationTimer;

  SafetyNotifier(this._repository, this._ref) : super(const SafetyState());

  @override
  void dispose() {
    _periodicLocationTimer?.cancel();
    super.dispose();
  }

  /// Trigger SOS Alert with emergency contact broadcast
  Future<SosTriggerResult?> triggerSos({
    bool autoDialHelpline = true,
    BuildContext? context,
  }) async {
    state = state.copyWith(isSendingSos: true, errorMessage: null);

    final authState = _ref.read(authControllerProvider);
    final user = authState is Authenticated ? authState.user : null;
    final userName = user?.displayName ?? 'Sanchari Traveler';
    final userPhone = user?.phone ?? '+919876543210';
    final contacts = user?.emergencyContacts ?? const [
      EmergencyContact(name: 'Emergency Services', phone: '112', relation: 'other', isPrimary: true),
    ];

    final result = await _repository.triggerSos(
      userName: userName,
      userPhone: userPhone,
      location: state.currentLocation,
      emergencyContacts: contacts,
      autoDialNumber: '112',
    );

    return result.when(
      success: (data) {
        state = state.copyWith(
          isSosActive: true,
          isLocationSharingActive: true,
          isSendingSos: false,
          activeSessionId: data.sessionId,
          publicTrackingUrl: data.publicTrackingUrl,
          lastSosResult: data,
        );

        // Start 10-second continuous streaming
        _startPeriodicLocationStreaming(data.sessionId);

        // Optionally dial national emergency helpline (112)
        if (autoDialHelpline) {
          _dialEmergencyNumber(data.emergencyHelpline);
        }

        return data;
      },
      failure: (err) {
        state = state.copyWith(
          isSendingSos: false,
          errorMessage: err.message,
        );
        return null;
      },
    );
  }

  /// Start Live Trip Location Sharing
  Future<bool> startTripShare() async {
    final authState = _ref.read(authControllerProvider);
    final user = authState is Authenticated ? authState.user : null;
    final userName = user?.displayName ?? 'Sanchari Traveler';
    final userPhone = user?.phone ?? '+919876543210';
    final contacts = user?.emergencyContacts ?? [];

    final result = await _repository.startTripShare(
      userName: userName,
      userPhone: userPhone,
      location: state.currentLocation,
      emergencyContacts: contacts,
      durationHours: 12,
    );

    return result.when(
      success: (data) {
        final sessionId = data['sessionId']?.toString() ?? 'trip_${Date.now()}';
        final trackingUrl = data['publicTrackingUrl']?.toString() ?? 'http://localhost:5000/live-track/$sessionId';

        state = state.copyWith(
          isLocationSharingActive: true,
          isSosActive: false,
          activeSessionId: sessionId,
          publicTrackingUrl: trackingUrl,
        );

        _startPeriodicLocationStreaming(sessionId);
        return true;
      },
      failure: (_) => false,
    );
  }

  /// Stop active sharing / SOS session
  Future<void> stopSharing() async {
    _periodicLocationTimer?.cancel();
    _periodicLocationTimer = null;

    if (state.activeSessionId != null) {
      await _repository.stopSession(state.activeSessionId!);
    }

    state = state.copyWith(
      isLocationSharingActive: false,
      isSosActive: false,
      activeSessionId: null,
      publicTrackingUrl: null,
    );
  }

  /// Periodic location streaming every 10 seconds
  void _startPeriodicLocationStreaming(String sessionId) {
    _periodicLocationTimer?.cancel();
    _periodicLocationTimer = Timer.periodic(const Duration(seconds: 10), (_) async {
      // Simulate minor GPS movement for realistic tracking simulation
      final newLat = state.currentLocation.lat + (0.00015 * (0.5 - (DateTime.now().second % 3) * 0.3));
      final newLng = state.currentLocation.lng + (0.00018 * (0.5 - (DateTime.now().second % 2) * 0.3));

      final updatedLoc = LiveLocationData(
        lat: newLat,
        lng: newLng,
        accuracy: 6.0,
        speed: (state.currentLocation.speed + 2) % 35,
        battery: 87,
        address: state.currentLocation.address,
      );

      state = state.copyWith(currentLocation: updatedLoc);

      await _repository.updateLiveLocation(
        sessionId: sessionId,
        lat: updatedLoc.lat,
        lng: updatedLoc.lng,
        speed: updatedLoc.speed,
        battery: updatedLoc.battery,
        address: updatedLoc.address,
      );
    });
  }

  Future<void> _dialEmergencyNumber(String number) async {
    final telUrl = Uri.parse('tel:$number');
    if (await canLaunchUrl(telUrl)) {
      await launchUrl(telUrl);
    }
  }
}
